External/SPMmodified/xASL/xASL_fileparts.m

```diff
@@ -1,63 +1,69 @@
-function [Fpath, Ffile, Fext] = xASL_fileparts(InputPath)
-% Wrapper around the fileparts.m that treats nii.gz as an extension
-% (this also works with other extensions)
+function [Fpath, Ffile, Fext, SuffixSPM] = xASL_fileparts(InputPath)
+%xASL_fileparts Wrapper around fileparts for special extensions
 % FORMAT: [Fpath, Ffile, Fext] = xASL_fileparts(InputPath)
 %
 % INPUT:
-%   InputPath - input file name with path, ending either .ext or .ext.gz
+%   InputPath - input file name with path, ending either .ext(.gz) or .ext(.gz),n
 % OUTPUT:
 %   Fpath     - path or empty string
 %   Ffile     - file name
 %   Fext      - file extension or empty string
+%   SuffixSPM - volume extension for NIfTI files (per spm usage)
 % -----------------------------------------------------------------------------------------------------------------------------------------------------
 % DESCRIPTION: Returns the path, file name, and file extension for InputPath using the fileparts.m function.
 %              If a file ending at nii.gz is given, then the whole nii.gz is returned as the extension.
-%              Does not verify the existence of the file, or existence of .nii or .nii.gz
-% EXAMPLE: xASL_fileparts('/path/file.nii');
-%          xASL_fileparts('/path/file.nii.gz','file');
-%          xASL_fileparts('c:\path\file.nii');
+%              Does not verify the existence of the file, or existence of .nii or .nii.gz.
+%              This function uses the following steps:
+%              1. Catch SPM suffixes
+%              2. Manage .gz (double) extensions  
+%              3. Manage folders differently than files
 % -----------------------------------------------------------------------------------------------------------------------------------------------------
-%
+% EXAMPLES: ['/path' 'file' '.nii']         = xASL_fileparts('/path/file.nii');
+%          ['/path' 'file' '.nii.gz']      = xASL_fileparts('/path/file.nii.gz');
+%          ['/path' 'file' '.nii.gz' ',2'] = xASL_fileparts('/path/file.nii.gz,2');
+%          ['c:\path' 'file' '.nii']       = xASL_fileparts('c:\path\file.nii') (on Windows)
 % __________________________________
-% Copyright (R) 2015-2019 ExploreASL
-%
-% 2015-01-01 HJ
+% Copyright 2015-2024 ExploreASL
 
+    %% Admin
     % First catch cell
     if iscell(InputPath)
         if length(InputPath)>1
             warning('InputPath contained multiple cells, using the first only');
         end
         InputPath = InputPath{1};
     end
-        
-    % Now catch SPM suffixes
-    [Ind1, Ind2] = regexp(InputPath,',\d');
+
+    %% 1. Catch SPM suffixes
+    [Fpath, Ffile, Fext] = fileparts(InputPath);
+
+    [Ind1, Ind2] = regexp(Fext, ',\d+');
```

Comment:
To avoid bugs, best to look at endings only
',\d+$'
otherwise, it could find this in the middle of the path for weird cases like this... "/tmp/test,1Henk/file.nii"

if doing this, you can skip Ind2

`Ind1 = regexp(Fext, ',\d+$');`

Reply:
>> [fPath, fFile, fExt] = fileparts('/tmp/test,1Henk/file.nii')

fPath =

    '/tmp/test,1Henk'


fFile =

    'file'


fExt =

    '.nii'

I don't mind replacing this, but the Fext already has the end. It could still be any weird thing right?
E.g., we would catch fExt = '.nii,1.gz'
Anyway, I don't mind much here :)

Reply:
You are right. That example above will work. But I meant a weird case, but that is not likely to happen but in theory it could.

Like
[Fpath, Ffile, Fext] = fileparts('perfectlynormalifile/savedbyHenk.M.M.Mutsaerts,1April-withoutextension')
For that, we would loose a large part of the file as a supposedly SPM extension.

So my correction would be able to handle this without affecting the other functionality.

Reply:
[Fpath, Ffile, Fext] = fileparts('perfectlynormalifile/savedbyHenk.M.M.Mutsaerts,1April-withoutextension')

Fpath =

    'perfectlynormalifile'


Ffile =

    'savedbyHenk.M.M'


Fext =

    '.Mutsaerts,1April-withoutextension'

my solution:


[Ind1, Ind2] = regexp(Fext, ',\d+')

[Ind1, Ind2] = regexp(Fext, ',\d+')

Ind1 =

    11


Ind2 =

    12

Your solution:

Ind1 = regexp(Fext, ',\d+$')

Ind1 =

     []


But the whole point here is to fix an extension. So if the extension is incorrect, the whole code is moot. This code only happens for NIfTIs, where they obligatory end with .nii(.gz). But I'm still happy to change to your code, it will still function the same :)

Probably your code is better, issuing an error when the catched SPM extension is empty.

Reply:
Yes, you are of course right in your observation with the current code. Because I haven't described the proposed change fully. 
To make my change fully working, you of course remove the condition
``if isempty(Index1)
        error('Something wrong with this file extension');``

If Index1 is empty, then SuffixSPM = ''.  If not empty, we assign the suffix.
This way, we make it work with nii.gz, we make it work with weird filenames "fileblabla,1April, blabla.nii.gz" etc.

Shall I make this edit, or you want to?

Reply:
No, I disagree, because there should not be a comma in the filename. That is why I wrote we should revert to what I had. So your solution is technically better, but content-wise not:

1. A filename with a comma is illegal (but this is for the BIDS validity check, not for this function)
2. A filename with a comma needs to be dealt with

Reply:
I'm confused. If my solution is technically better, why not simply using it? What's wrong with having a better solution? 

Currently, it will crash for all files with commas. 
I understand that with BIDS, this should not happen. But that function should work universally also for non-BIDS files, DICOM files and everything. So we're writing something that is nice for BIDS,but crashing with anything else?

Reply:
True, xASL_fileparts is designed for NIfTIs only.
So it should first run
1. [folder, filename, extension] = fileparts(pathInput);

2. folder can contain commas or whatever.

3. If we detect a NIfTI file (extension contains .nii) then we continue (otherwise we skip our addition)

4. filename cannot contain a comma (but it doesn't matter for this code, it will be catched elsewhere, and we don't change the filename with this code)

5. Now we're left with the extension, if we detect .nii that is followed by a comma, we remove the comma

Reply:
Henk, please read the discussion above. I know it's not easy for you, but try to change your opinion once.

We have a working solution. This works perfectly with commas in foldername. It works perfectly with commas in filename (and yes, these are allowed and filenames do include commas sometimes).

And now, you are argumenting that we stop using code that works and use a different, much more complex code that detects NIfTI, changes execution etc and in the end does the same.

It makes absolutely no sense. But if you want, please do the changes. But just make sure they work.